settings: classify public vs non-public cluster settings

knz opened the pull request:
First commit from #42563.

This PR makes three user-visible changes recommended/requested by @johnrk: 

- it annotates all the cluster settings into broad public (= documented+supported) and non-public (= non-documented/non-supported) categories.
- the categories are now displayed alongside the settings in the output of `SHOW ALL CLUSTER SETTINGS`
- a new statement `SHOW CLUSTER SETTINGS` is introduced that lists only the public settings.

By default during implementation, a setting starts as non-public.

New settings should only be promoted to the public category after the team has built confidence in a broad understanding of all its possible effects, and after starting a doc project to explain the setting.

In this PR, I have manually opted in the following settings to become public:

```
                          variable                         |                    desc
+----------------------------------------------------------+---------------------------------------------+
  cloudstorage.gs.default.key                              | if set, JSON key to use during Google Cl...
  cloudstorage.http.custom_ca                              | custom root CA (appended to system's def...
  cloudstorage.timeout                                     | the timeout for import/export storage op...
  cluster.organization                                     | organization name...
  cluster.preserve_downgrade_option                        | disable (automatic or manual) cluster ve...
  debug.panic_on_failed_assertions                         | panic when an assertion fails rather tha...
  diagnostics.forced_stat_reset.interval                   | interval after which pending diagnostics...
  diagnostics.reporting.enabled                            | enable reporting diagnostic metrics to c...
  diagnostics.reporting.interval                           | interval at which diagnostics data shoul...
  external.graphite.endpoint                               | if nonempty, push server metrics to the ...
  external.graphite.interval                               | the interval at which metrics are pushed...
  kv.allocator.load_based_lease_rebalancing.enabled        | set to enable rebalancing of range lease...
  kv.allocator.load_based_rebalancing                      | whether to rebalance based on the distri...
  kv.allocator.qps_rebalance_threshold                     | minimum fraction away from the mean a st...
  kv.allocator.range_rebalance_threshold                   | minimum fraction away from the mean a st...
  kv.bulk_io_write.max_rate                                | the rate limit (bytes/sec) to use for wr...
  kv.closed_timestamp.follower_reads_enabled               | allow (all) replicas to serve consistent...
  kv.range_split.by_load_enabled                           | allow automatic splits of ranges based o...
  kv.range_split.load_qps_threshold                        | the QPS over which, the range becomes a ...
  kv.rangefeed.enabled                                     | if set, rangefeed registration is enable...
  kv.replication_reports.interval                          | the frequency for generating the replica...
  kv.snapshot_rebalance.max_rate                           | the rate limit (bytes/sec) to use for re...
  kv.snapshot_recovery.max_rate                            | the rate limit (bytes/sec) to use for re...
  kv.transaction.max_intents_bytes                         | maximum number of bytes used to track wr...
  kv.transaction.max_refresh_spans_bytes                   | maximum number of bytes used to track re...
  server.eventlog.ttl                                      | if nonzero, event log entries older than...
  server.host_based_authentication.configuration           | host-based authentication configuration ...
  server.rangelog.ttl                                      | if nonzero, range log entries older than...
  server.remote_debugging.mode                             | set to enable remote debugging, localhos...
  server.time_until_store_dead                             | the time after which if there is no new ...
  server.web_session_timeout                               | the duration that a newly created web se...
  sql.defaults.default_int_size                            | the size, in bytes, of an INT type...
  sql.defaults.distsql                                     | default distributed SQL execution mode [...
  sql.defaults.results_buffer.size                         | default size of the buffer that accumula...
  sql.defaults.serial_normalization                        | default handling of SERIAL in table defi...
  sql.distsql.max_running_flows                            | maximum number of concurrent flows that ...
  sql.distsql.temp_storage.joins                           | set to true to enable use of disk for di...
  sql.distsql.temp_storage.sorts                           | set to true to enable use of disk for di...
  sql.metrics.statement_details.dump_to_logs               | dump collected statement statistics to n...
  sql.metrics.statement_details.enabled                    | collect per-statement query statistics...
  sql.metrics.statement_details.plan_collection.enabled    | periodically save a logical plan for eac...
  sql.metrics.statement_details.plan_collection.period     | the time until a new logical plan is col...
  sql.metrics.statement_details.threshold                  | minimum execution time to cause statisti...
  sql.metrics.transaction_details.enabled                  | collect per-application transaction stat...
  sql.stats.automatic_collection.enabled                   | automatic statistics collection mode...
  sql.stats.histogram_collection.enabled                   | histogram collection mode...
  sql.stats.post_events.enabled                            | if set, an event is logged for every CRE...
  sql.trace.log_statement_execute                          | set to true to enable logging of execute...
  sql.trace.session_eventlog.enabled                       | set to true to enable session tracing. N...
  sql.trace.txn.enable_threshold                           | duration beyond which all transactions a...
  timeseries.storage.enabled                               | if set, periodic timeseries data is stor...
  timeseries.storage.resolution_10s.ttl                    | the maximum age of time series data stor...
  timeseries.storage.resolution_30m.ttl                    | the maximum age of time series data stor...
  trace.debug.enable                                       | if set, traces for recent requests can b...
  trace.lightstep.token                                    | if set, traces go to Lightstep using thi...
  trace.zipkin.collector                                   | if set, traces go to the given Zipkin in...
  version                                                  | set the active cluster version in the fo...
```

knz committed:
settings: clarify the "non reportability" property

Prior to this patch, "confidential" settings were masked by removing
them from the list of settings returned by `settings.Keys()` (using a
`hidden` bit), so that all the reporting functions would omit them:
SHOW ALL CLUSTER SETTINGS and the admin UI RPC to list all settings.

This was inadequate, because this prevents a user to even discover the
confidential setting exists in the first place. Worse even the setting
would be even omitted from the table of all settings included in docs!

This patch improves the situation by revealing the name of
confidential settings in reports, leaving only their value
hidden ("`<redacted>`"). The "confidentiality" is renamed to
"reportability": a non-reportable setting is one whose value cannot be
included in reports.

To achieve this, the `Lookup` function now takes an extra argument
with the purpose of the lookup; when the purpose is a report, it hides
non-reportable settings by encapsulating them dynamically in an object
of type `MaskedSetting`. This advertises the same type and description
but erases the `String()` representation of its underlying setting
object.

Meanwhile, _deprecated_ settings were also using the same "hidden"
bit, but for an entirely unrelated reasons: for those we actually want
to omit them from advertised lists of available settings. For these,
this patch keeps the same mechanism, albeit renaming the "hidden" flag
to `isDeprecated`, to clarify its purpose.

Release note (general change): CockroachDB will now advertise some
previously-hidden cluster settings, such as `enterprise.license`, in
reports, such as generated by SHOW ALL CLUSTER SETTINGS. Only the
names are listed; the values are still redacted out. The values
can be accessed/modified using the specific statements SET/SHOW
CLUSTER SETTING (singular).

knz committed:
settings,*: separate settings into categories: public and reserved

Previously, all settings were listed side-by-side and it wasn't clear
which were those for which the CockroachDB team provides documentation
and support, and those which are not meant to be customized without
supervision.

This created friction with users, who routinely explore, discover the
settings anyway, and are interested to know which knobs are truly
tunable.

This patch improves upon this situation by classifying all settings
into the following 2 categories:

- **Public** indicates that a setting is documented, the range of
  possible values yields predictable results, and the CockroachDB team
  is there to assist if issues occur as a result of the customization.

  In short: "Go ahead but be careful."

- **Reserved** - which is the default when creating new settings -
  indicates that a setting is not documented and the CockroachDB team
  has not yet developed internal experience about the impact of
  customizing it to other values.

  In short: "Use at your own risk."

This status is also reported in the `crdb_internal.cluster_settings`
vtable and the output of SHOW ALL CLUSTER SETTINGS.

The specific cluster settings that become "Public" in this change:

```
                        variable                        |                    desc
+-------------------------------------------------------+---------------------------------------------+
  cloudstorage.gs.default.key                           | if set, JSON key to use during Google Cl...
  cloudstorage.http.custom_ca                           | custom root CA (appended to system's def...
  cloudstorage.timeout                                  | the timeout for import/export storage op...
  cluster.organization                                  | organization name...
  cluster.preserve_downgrade_option                     | disable (automatic or manual) cluster ve...
  diagnostics.forced_stat_reset.interval                | interval after which pending diagnostics...
  diagnostics.reporting.enabled                         | enable reporting diagnostic metrics to c...
  diagnostics.reporting.interval                        | interval at which diagnostics data shoul...
  enterprise.license                                    | the encoded cluster license...
  external.graphite.endpoint                            | if nonempty, push server metrics to the ...
  external.graphite.interval                            | the interval at which metrics are pushed...
  kv.allocator.load_based_lease_rebalancing.enabled     | set to enable rebalancing of range lease...
  kv.allocator.load_based_rebalancing                   | whether to rebalance based on the distri...
  kv.allocator.qps_rebalance_threshold                  | minimum fraction away from the mean a st...
  kv.allocator.range_rebalance_threshold                | minimum fraction away from the mean a st...
  kv.bulk_io_write.max_rate                             | the rate limit (bytes/sec) to use for wr...
  kv.closed_timestamp.follower_reads_enabled            | allow (all) replicas to serve consistent...
  kv.rangefeed.enabled                                  | if set, rangefeed registration is enable...
  kv.replication_reports.interval                       | the frequency for generating the replica...
  kv.snapshot_rebalance.max_rate                        | the rate limit (bytes/sec) to use for re...
  kv.snapshot_recovery.max_rate                         | the rate limit (bytes/sec) to use for re...
  kv.transaction.max_intents_bytes                      | maximum number of bytes used to track wr...
  kv.transaction.max_refresh_spans_bytes                | maximum number of bytes used to track re...
  server.clock.forward_jump_check_enabled               | if enabled, forward clock jumps > max_of...
  server.clock.persist_upper_bound_interval             | the interval between persisting the wall...
  server.eventlog.ttl                                   | if nonzero, event log entries older than...
  server.host_based_authentication.configuration        | host-based authentication configuration ...
  server.rangelog.ttl                                   | if nonzero, range log entries older than...
  server.remote_debugging.mode                          | set to enable remote debugging, localhos...
  server.shutdown.drain_wait                            | the amount of time a server waits in an ...
  server.shutdown.query_wait                            | the server will wait for at least this a...
  server.time_until_store_dead                          | the time after which if there is no new ...
  server.web_session_timeout                            | the duration that a newly created web se...
  sql.defaults.default_int_size                         | the size, in bytes, of an INT type...
  sql.defaults.results_buffer.size                      | default size of the buffer that accumula...
  sql.defaults.serial_normalization                     | default handling of SERIAL in table defi...
  sql.distsql.max_running_flows                         | maximum number of concurrent flows that ...
  sql.distsql.temp_storage.joins                        | set to true to enable use of disk for di...
  sql.distsql.temp_storage.sorts                        | set to true to enable use of disk for di...
  sql.metrics.statement_details.dump_to_logs            | dump collected statement statistics to n...
  sql.metrics.statement_details.enabled                 | collect per-statement query statistics...
  sql.metrics.statement_details.plan_collection.enabled | periodically save a logical plan for eac...
  sql.metrics.statement_details.plan_collection.period  | the time until a new logical plan is col...
  sql.metrics.statement_details.threshold               | minimum execution time to cause statisti...
  sql.metrics.transaction_details.enabled               | collect per-application transaction stat...
  sql.stats.automatic_collection.enabled                | automatic statistics collection mode...
  sql.stats.automatic_collection.fraction_stale_rows    | target fraction of stale rows per table ...
  sql.stats.automatic_collection.min_stale_rows         | target minimum number of stale rows per ...
  sql.stats.histogram_collection.enabled                | histogram collection mode...
  sql.stats.post_events.enabled                         | if set, an event is logged for every CRE...
  sql.trace.log_statement_execute                       | set to true to enable logging of execute...
  sql.trace.session_eventlog.enabled                    | set to true to enable session tracing. N...
  sql.trace.txn.enable_threshold                        | duration beyond which all transactions a...
  timeseries.storage.enabled                            | if set, periodic timeseries data is stor...
  timeseries.storage.resolution_10s.ttl                 | the maximum age of time series data stor...
  timeseries.storage.resolution_30m.ttl                 | the maximum age of time series data stor...
  trace.debug.enable                                    | if set, traces for recent requests can b...
  trace.lightstep.token                                 | if set, traces go to Lightstep using thi...
  trace.zipkin.collector                                | if set, traces go to the given Zipkin in...
  version                                               | set the active cluster version in the fo...
```

Release note (general change): It is now possible to easily identify
cluster settings for which tuning effects are known and documented,
via the new column `public` in the output of `SHOW ALL CLUSTER
SETTINGS` and the virtual table `crdb_internal.cluster_settings`.

Release note (ui change): The cluster settings page now lists
public and reserved settings in two separate tables.

knz committed:
sql: new stmt SHOW [PUBLIC] CLUSTER SETTINGS

The previous commit included a distinction between public and
non-public cluster settings. Both are displayed in the output of `SHOW
ALL CLUSTER SETTINGS`, with a pair of columns that indicates the
visibility of each setting.

For convenience, this setting introduces the new statement variant
`SHOW [PUBLIC] CLUSTER SETTINGS` which filters the list to just
public settings and hides the visibility columns.

Release note (sql change): The new statement `SHOW PUBLIC CLUSTER
SETTINGS` (which can be abbreviated to just `SHOW CLUSTER SETTINGS`)
can be used to list just the public cluster settings that are
supported for tuning and configuration.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42520)
<!-- Reviewable:end -->

knz commented:
@johnrk is there a github issue filed about this already? (I haven't found one) If it exists, I'll need to care about closing it when this PR gets merged.

dt commented:
I'm a big fan of the overall change here. However I think there should just be two categories: public and internal: I'm not convinced that there should be a difference between `Restricted` and `Sensitive`. 

To me the contract on all settings started out as "don't mess with this unless you know what you're doing i.e. because someone at Cockroach Labs told you to or you've read the code/docs/etc and know what exactly what it will do". 

Over time we've increasingly made some settings more self-service: we've instructed users on when and how to set them, we've considered that end-users will set them and considered their effects, added guardrails, tested, etc These are what I think of as "public" or "user-facing" settings, and I love that this change is specifically identifying these. 

However, I think all the internal settings still share the same contract "this could do bad things if used incorrectly so don't touch it unless you know what you're doing". I think flagging some them with this explicitly but not others might give a false sense of security to the internal, but not sensitive, ones but also isn't particularly helpful at identifying _why_ or how a certain one is particularly scary. IMO every non-public setting should just be that -- non-public -- and if a certain setting can cause a certain problem, that specific concern should just be documented.

I've argued against the "sensitive" or "hidden" settings in the past as well and still have the same objections to them: I also dislike hiding them since, if they really can have such insidious effects, it could be _really helpful_ to know what they're set to. If someone else on your team _did_ set one and you see a script or mention in a log/chat or whatever, having it not show up in docs/introspection/etc will just add to confusion. With the explicit partitioning of public settings into their own, special category, I think there's even less reason to keep the current "Sensitive" pattern around. (I also think the language itself is a big off-putting/patronizing -- I don't like being told I need "supervision".)

Anyway, like I said, I _really like_ the gist of this change.

knz commented:
@dt no problem for me to remove the sensitive bit. I was keeping there because the distinction was there before and I wasn't confident about removing it. But your argument convinces me it's a good thing.

The "hidden" bit is no more, see the first commit for a clarification.

dt commented on pkg/util/log/crash_reporting.go:
```diff
@@ -58,21 +58,21 @@ var (
 
 	// CrashReports wraps "diagnostics.reporting.send_crash_reports".
-	CrashReports = settings.RegisterBoolSetting(
+	CrashReports = settings.RegisterPublicBoolSetting(
```
Since crash reporting *also* obeys the public diagnostics setting, I don't think we need to muddy the waters by also making this one public (indeed, we could probably just remove it entirely, but that's separate change)

knz replied:
Done.

dt commented on pkg/settings/registry.go:
```diff
@@ -72,7 +72,7 @@ func register(key, desc string, s Setting) {
 	res = make([]string, 0, len(Registry))
 	for k := range Registry {
-		if Registry[k].Hidden() {
+		if !Registry[k].IsReportable() {
```
detailed more below, but I think we should remove this behavior.

knz replied:
Done.

dt commented on pkg/settings/setting.go:
```diff
@@ -197,17 +197,54 @@ type Setting interface {
 	getSlotIdx() int
-	Hidden() bool
+
+	// IsReportable indicates whether the setting can be *listed*
```
I realize you just renamed this and didn't introduce it here, but since we we are reworking/documenting the API here: I think we should re-define the behavior of this flag to be that _the value_ of this setting defaults to being hidden / redacted in listings / screenshots / telemetry, not that the setting itself is. I think it would be valuable to an operator and telemetry to know that a certain setting exists and/or is set, even if we think we should default to hiding the value from broad listings and telemetry.

From that perspective, I'm fine with the "Confidential" name, but also fine with "Reportable"

knz replied:
Done.

dt commented on pkg/sql/parser/sql.y:
```diff
@@ -3477,14 +3477,23 @@ show_csettings_stmt:
+    $$.val = &tree.ShowClusterSettingList{All: true}
   }
 | SHOW ALL CLUSTER error // SHOW HELP: SHOW CLUSTER SETTING
+| SHOW CLUSTER SETTINGS
```
would it make sense, for completeness, to have a `SHOW INTERNAL CLUSTER SETTINGS` counterpart to `SHOW PUBLIC CLUSTER SETTINGS`?

knz replied:
I am not too fond of it. Specialized syntactic variants are intended for frequently-used cases. This is (hopefully) a rare case, which has a separate syntax already: `select * from [show all cluster settings] where public is false`.

knz reviewed (commented):
I have amended the first commit as suggested: all non-deprecated settings _names_ are now visible in reports, and only the values are redacted.

I also amended the 2nd commit to remove the reserved/sensitive distinction. All previously-sensitive settings are now reserved (non-public).

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42520)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @johnrk, and @knz)

___
*[pkg/sql/stats/automatic_stats.go, line 67 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Ltv7Fj5-Zsd-OAyQIIq:-Ltv7Fj6BVhaZOtVkEVd:bcr1xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/efa2de6498b5575eb74a338312af534bad4c8cb0/pkg/sql/stats/automatic_stats.go#L67)):*
> ```Go
> // statistics on that table are refreshed, in addition to the constant value
> // AutomaticStatisticsMinStaleRows.
> var AutomaticStatisticsFractionStaleRows = settings.RegisterNonNegativeFloatSetting(
> ```

I think this setting and the one below should be public since both are well documented as knobs for adjusting the frequency of stats refreshes.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42520)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @johnrk, and @rytaft)

___
*[pkg/sql/stats/automatic_stats.go, line 67 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Ltv7Fj5-Zsd-OAyQIIq:-LtvAoLj4OmsTO1vlONb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/efa2de6498b5575eb74a338312af534bad4c8cb0/pkg/sql/stats/automatic_stats.go#L67)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this setting and the one below should be public since both are well documented as knobs for adjusting the frequency of stats refreshes.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 60 files at r5, 1 of 13 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42520)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @johnrk, and @knz)

___
*[pkg/sql/stats/automatic_stats.go, line 67 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Ltv7Fj5-Zsd-OAyQIIq:-LtvHc7dBnUcfsv7v_kF:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/efa2de6498b5575eb74a338312af534bad4c8cb0/pkg/sql/stats/automatic_stats.go#L67)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->

johnrk commented:
> @johnrk is there a github issue filed about this already? (I haven't found one) If it exists, I'll need to care about closing it when this PR gets merged.

@knz , thanks for jumping on this. There is no github issue filed.

johnrk-zz reviewed (approved):
lgtm

bdarnell reviewed (commented):
LGTM

I left a lot of comments about settings that I think should move from one category to the other. There's probably still room to move even more to the non-public category. 

> Reviewed 5 of 68 files at r4, 54 of 60 files at r5, 2 of 13 files at r7, 1 of 1 files at r8, 10 of 10 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-:-Lu-826v2Z5-FFnUUCAJ:brn6p4i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @johnrk, and @knz)

___
*[docs/generated/settings/settings.html, line 9 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu--zVn8_5NGa0QPFFQ:-Lu--zVn8_5NGa0QPFFR:b-yhkuf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L9)):*
> ```HTML
> <tr><td><code>cluster.organization</code></td><td>string</td><td><code></code></td><td>organization name</td></tr>
> <tr><td><code>cluster.preserve_downgrade_option</code></td><td>string</td><td><code></code></td><td>disable (automatic or manual) cluster version upgrade from the specified version until reset</td></tr>
> <tr><td><code>cluster.secret</code></td><td>string</td><td><code><random></code></td><td>cluster specific secret</td></tr>
> ```

`cluster.secret` shouldn't be public. 
___
*[docs/generated/settings/settings.html, line 10 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-02ArAwsSjnQXYVnY:-Lu-02ArAwsSjnQXYVnZ:b-47m0p8) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L10)):*
> ```HTML
> <tr><td><code>cluster.preserve_downgrade_option</code></td><td>string</td><td><code></code></td><td>disable (automatic or manual) cluster version upgrade from the specified version until reset</td></tr>
> <tr><td><code>cluster.secret</code></td><td>string</td><td><code><random></code></td><td>cluster specific secret</td></tr>
> <tr><td><code>debug.panic_on_failed_assertions</code></td><td>boolean</td><td><code>false</code></td><td>panic when an assertion fails rather than reporting</td></tr>
> ```

`debug.panic_on_failed_assertions` probably shouldn't be public. 
___
*[docs/generated/settings/settings.html, line 23 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-18ad-ltlJNwtVyQq:-Lu-18ad-ltlJNwtVyQr:b-pwjs0) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L23)):*
> ```HTML
> <tr><td><code>kv.bulk_io_write.max_rate</code></td><td>byte size</td><td><code>1.0 TiB</code></td><td>the rate limit (bytes/sec) to use for writes to disk on behalf of bulk io ops</td></tr>
> <tr><td><code>kv.closed_timestamp.follower_reads_enabled</code></td><td>boolean</td><td><code>true</code></td><td>allow (all) replicas to serve consistent historical reads based on closed timestamp information</td></tr>
> <tr><td><code>kv.range_split.by_load_enabled</code></td><td>boolean</td><td><code>true</code></td><td>allow automatic splits of ranges based on where load is concentrated</td></tr>
> ```

`kv.range_split.by_load_enabled` and `kv.range_split.load_qps_threshold` should probably not be public. 
___
*[docs/generated/settings/settings.html, line 38 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-0tF77o_fWevA-UFA:-Lu-0tF85FHIJoQ6JKyJ:bl5jjx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L38)):*
> ```HTML
> <tr><td><code>server.web_session_timeout</code></td><td>duration</td><td><code>168h0m0s</code></td><td>the duration that a newly created web session will be valid</td></tr>
> <tr><td><code>sql.defaults.default_int_size</code></td><td>integer</td><td><code>8</code></td><td>the size, in bytes, of an INT type</td></tr>
> <tr><td><code>sql.defaults.distsql</code></td><td>enumeration</td><td><code>auto</code></td><td>default distributed SQL execution mode [off = 0, auto = 1, on = 2]</td></tr>
> ```

Should `sql.defaults.distsql` still be public?
___
*[docs/generated/settings/settings.html, line 79 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-0cKf6PuJ-JpmtzkV:-Lu-0cKf6PuJ-JpmtzkW:b-5ucn74) ([raw file](https://github.com/cockroachdb/cockroach/blob/09b391536f72349c3951ce6b75c5231b58933b07/docs/generated/settings/settings.html#L79)):*
> ```HTML
> <tr><td><code>schemachanger.lease.duration</code></td><td>duration</td><td><code>5m0s</code></td><td>the duration of a schema change lease</td></tr>
> <tr><td><code>schemachanger.lease.renew_fraction</code></td><td>float</td><td><code>0.5</code></td><td>the fraction of schemachanger.lease_duration remaining to trigger a renew of the lease</td></tr>
> <tr><td><code>server.clock.forward_jump_check_enabled</code></td><td>boolean</td><td><code>false</code></td><td>if enabled, forward clock jumps > max_offset/2 will cause a panic</td></tr>
> ```

`server.clock.forward_jump_check_enabled` and `server.clock.persist_upper_bound_interval` should be public. 
___
*[docs/generated/settings/settings.html, line 91 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-0nHIAl9nG2QpUabr:-Lu-0nHIAl9nG2QpUabs:b-sa6lxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/09b391536f72349c3951ce6b75c5231b58933b07/docs/generated/settings/settings.html#L91)):*
> ```HTML
> <tr><td><code>server.rangelog.ttl</code></td><td>duration</td><td><code>720h0m0s</code></td><td>if nonzero, range log entries older than this duration are deleted every 10m0s. Should not be lowered below 24 hours.</td></tr>
> <tr><td><code>server.remote_debugging.mode</code></td><td>string</td><td><code>local</code></td><td>set to enable remote debugging, localhost-only or disable (any, local, off)</td></tr>
> <tr><td><code>server.shutdown.drain_wait</code></td><td>duration</td><td><code>0s</code></td><td>the amount of time a server waits in an unready state before proceeding with the rest of the shutdown process</td></tr>
> ```

`server.shutdown.drain_wait` and `server.shutdown.query_wait` should probably be public. 
___
*[pkg/settings/registry.go, line 90 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-2wde8ZinR4pnGxz1:-Lu-2wde8ZinR4pnGxz2:b-yyl32u) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/settings/registry.go#L90)):*
> ```Go
> 	res = make([]string, 0, len(registry))
> 	for k := range registry {
> 		if registry[k].isDeprecated() {
> ```

I wouldn't expect deprecation to hide a setting from enumeration. The main effect of deprecation should be to issue a warning when the setting is changed; otherwise it should work as usual. 
___
*[pkg/settings/string.go, line 106 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-6WzVEqq7LjNS71V2:-Lu-6WzVEqq7LjNS71V3:b-s7afq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/settings/string.go#L106)):*
> ```Go
> 		validateFn:   validateFn,
> 	}
> 	// By defauult all string settings are considered to perhaps contain
> ```

s/uu/u/
___
*[pkg/sql/logictest/testdata/logic_test/show_source, line 86 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-7KDy-tfY0QwgTbp7:-Lu-7KDy-tfY0QwgTbp8:b-lo7wa5) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/sql/logictest/testdata/logic_test/show_source#L86)):*
> ```
> ----
> variable              value  setting_type  public  description
> cluster.organization  <redacted>      s             true    organization name
> ```

Redaction should only apply to telemetry reports; if a user (with suitable permissions; maybe this is admin-only), they should be able to see the real value here. 
___
*[pkg/sql/parser/sql.y, line 9506 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-7VbiBpjqzpJGIh_2:-Lu-7VbiBpjqzpJGIh_3:bxao766) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/sql/parser/sql.y#L9506)):*
> ```Yacc
> | PREPARE
> | PRIORITY
> | PUBLIC
> ```

Is there any concern with this keyword interacting with the `public` schema or is it OK since it's unreserved?


<!-- Sent from Reviewable.io -->

knz committed:
settings: rename "deprecated" to "retired"

The word "deprecated" in the context of the `settings` package truly
means that a setting has been obsoleted and already replaced by a
newer one, to the point that changes to the "deprecatd" settings are
ineffectual.

This conflicts with other uses of the word "deprecated" in other parts
of the project, where "deprecated" means "still in use, but _will_ be
replaced in the future".

To align the vocabulary, this commit renames the setting attribute to
"retired".

Release note: None

knz commented:
(Waiting for resolution of #42563 before merging this)

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42520)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, and @rytaft)

___
*[docs/generated/settings/settings.html, line 9 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu--zVn8_5NGa0QPFFQ:-Lu1yrTp07TiylRaxmbS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L9)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`cluster.secret` shouldn't be public. 
</blockquote></details>

Done.
___
*[docs/generated/settings/settings.html, line 10 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-02ArAwsSjnQXYVnY:-Lu1yxcQ9mMvH-__QWkW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L10)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`debug.panic_on_failed_assertions` probably shouldn't be public. 
</blockquote></details>

Done.
___
*[docs/generated/settings/settings.html, line 23 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-18ad-ltlJNwtVyQq:-Lu1z6H871zpt5gv2F5O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L23)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`kv.range_split.by_load_enabled` and `kv.range_split.load_qps_threshold` should probably not be public. 
</blockquote></details>

Done.
___
*[docs/generated/settings/settings.html, line 38 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-0tF77o_fWevA-UFA:-Lu1zN3_EKpbOIf8aOso:bc8k5x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/docs/generated/settings/settings.html#L38)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should `sql.defaults.distsql` still be public?
</blockquote></details>

Probably not. Removed.
___
*[docs/generated/settings/settings.html, line 79 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-0cKf6PuJ-JpmtzkV:-Lu1zEsDDeiFP8Bb-Ie4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09b391536f72349c3951ce6b75c5231b58933b07/docs/generated/settings/settings.html#L79)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`server.clock.forward_jump_check_enabled` and `server.clock.persist_upper_bound_interval` should be public. 
</blockquote></details>

Done.
___
*[docs/generated/settings/settings.html, line 91 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-0nHIAl9nG2QpUabr:-Lu1zKpi0qYPrW8sct2t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09b391536f72349c3951ce6b75c5231b58933b07/docs/generated/settings/settings.html#L91)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`server.shutdown.drain_wait` and `server.shutdown.query_wait` should probably be public. 
</blockquote></details>

Done.
___
*[pkg/settings/registry.go, line 90 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-2wde8ZinR4pnGxz1:-Lu1z_Ag-CshwTK8c0-O:b-d1w21i) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/settings/registry.go#L90)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I wouldn't expect deprecation to hide a setting from enumeration. The main effect of deprecation should be to issue a warning when the setting is changed; otherwise it should work as usual. 
</blockquote></details>

The word is/was incorrect. In the context of settings, the qualifier is really "retired" not "deprecated". Changing such a setting has no effect because a newer setting has taken over. It only remains supported for the purpose of reading its previously-set value. I'll rename.
___
*[pkg/settings/string.go, line 106 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-6WzVEqq7LjNS71V2:-Lu24ncb5Ti4n1U4k_2Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/settings/string.go#L106)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/uu/u/
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/show_source, line 86 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-7KDy-tfY0QwgTbp7:-Lu24oMf5pvwK_Pq1BL9:b-j6ujmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/sql/logictest/testdata/logic_test/show_source#L86)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Redaction should only apply to telemetry reports; if a user (with suitable permissions; maybe this is admin-only), they should be able to see the real value here. 
</blockquote></details>

Woah, good catch.
It's a bit more involved to achieve that, but while investigating I found a couple interesting things which needed fixing (separate PR #42563)
But it's now fixed.
___
*[pkg/sql/parser/sql.y, line 9506 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/42520#-Lu-7VbiBpjqzpJGIh_2:-Lu2-ADC9Z2fY-c3jszx:b-pavcg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/17045dd15ce36cc3ccfebc022206cc76a0264115/pkg/sql/parser/sql.y#L9506)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is there any concern with this keyword interacting with the `public` schema or is it OK since it's unreserved?
</blockquote></details>

It's OK because it's unreserved.


<!-- Sent from Reviewable.io -->

knz commented:
TFYR!

bors r+
